From 05064f333fe75f4b8ec3361889217eaa6a43394b Mon Sep 17 00:00:00 2001
From: Jean-Philippe Brucker <jean-philippe@linaro.org>
Date: Tue, 15 Feb 2022 15:48:41 +0000
Subject: [PATCH 48/97] KVM: arm64: prevent migration while setting up pKVM

do_pkvm_init() install the hyp vectors with cpu_hyp_init_context() on a
single CPU to run some hypercalls. If we get migrated to a different
CPUs, any hypercall such as that in create_hyp_exec_mappings() will
fail. Prevent migration until cpu_hyp_init_context() has run on all
CPUs.

Signed-off-by: Jean-Philippe Brucker <jean-philippe@linaro.org>
----
Quick fix for now, but maybe we could instead install the vectors
properly from the start?  Not familiar enough with the code at the
moment.
---
 arch/arm64/kvm/arm.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/arch/arm64/kvm/arm.c b/arch/arm64/kvm/arm.c
index 3e29af151671..9b596045a98d 100644
--- a/arch/arm64/kvm/arm.c
+++ b/arch/arm64/kvm/arm.c
@@ -2267,6 +2267,13 @@ int kvm_arch_init(void *opaque)
 	if (err)
 		return err;
 
+	/*
+	 * If pKVM is enabled, we're about to install the vectors for this CPU
+	 * only. Ensure we keep executing on it until hyp_install_host_vector()
+	 * has run everywhere.
+	 */
+	migrate_disable();
+
 	if (!in_hyp_mode) {
 		err = init_hyp_mode();
 		if (err)
@@ -2291,6 +2298,8 @@ int kvm_arch_init(void *opaque)
 		}
 	}
 
+	migrate_enable();
+
 	if (is_protected_kvm_enabled()) {
 		kvm_info("Protected nVHE mode initialized successfully\n");
 	} else if (in_hyp_mode) {
@@ -2306,6 +2315,7 @@ int kvm_arch_init(void *opaque)
 	if (!in_hyp_mode)
 		teardown_hyp_mode();
 out_err:
+	migrate_enable();
 	return err;
 }
 
-- 
2.34.1

